includes/Checker/Checks/Abstract_File_Check.php

```diff
@@ -0,0 +1,227 @@
+<?php
+/**
+ * Class WordPress\Plugin_Check\Checker\Checks\Abstract_File_Check
+ *
+ * @package plugin-check
+ */
+
+namespace WordPress\Plugin_Check\Checker\Checks;
+
+use Exception;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use WordPress\Plugin_Check\Checker\Check_Context;
+use WordPress\Plugin_Check\Checker\Check_Result;
+use WordPress\Plugin_Check\Checker\Static_Check;
+
+/**
+ * Base class for a check that inspects the plugin's files and contents.
+ *
+ * @since n.e.x.t
+ */
+abstract class Abstract_File_Check implements Static_Check {
+
+	/**
+	 * Internal cache for plugin-specific file lists.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private static $file_list_cache = array();
+
+	/**
+	 * Internal cache for file contents.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private static $file_contents_cache = array();
+
+	/**
+	 * Amends the given result by running the check on the associated plugin.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 *
+	 * @throws Exception Thrown when the check fails with a critical error (unrelated to any errors detected as part of
+	 *                   the check).
+	 */
+	final public function run( Check_Result $result ) {
+		$files = self::get_files( $result->plugin() );
+		$this->check_files( $result, $files );
+	}
+
+	/**
+	 * Amends the given result by running the check on the given list of files.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 * @param array        $files  List of absolute file paths.
+	 *
+	 * @throws Exception Thrown when the check fails with a critical error (unrelated to any errors detected as part of
+	 *                   the check).
+	 */
+	abstract protected function check_files( Check_Result $result, array $files );
+
+	/**
+	 * Filters a given list of files to only contain those with specific extension.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array  $files     List of absolute file paths.
+	 * @param string $extension File extension to match.
+	 * @return array Filtered $files list.
+	 */
+	final protected static function filter_files_by_extension( array $files, $extension ) {
```

Comment:
Why does this need to be static?

Reply:
See my reply in https://github.com/10up/plugin-check/pull/175#discussion_r1205834776.